configs/components/puppet-versions.rb

```diff
@@ -7,86 +7,70 @@
   require 'rubygems'
 
-  # Return all available puppet gem versions (from rubygems.org) matching a range
-  def available_puppet_gem_versions(range)
+  def latest_puppet_gem(req)
     require 'gems'
 
     @_puppet_all ||= Gems.versions 'puppet'
     @_puppet_in_range ||= {}
 
-    @_puppet_in_range[range] ||= begin
-      requirement = Gem::Requirement.create(range.split(' '))
+    @_puppet_in_range[req] ||= begin
+      requirement = Gem::Requirement.create(req)
 
       @_puppet_all.collect do |v|
         pupver = Gem::Version.new(v['number'])
         next unless requirement.satisfied_by?(pupver)
+
         pupver
-      end.compact.uniq.sort
+      end.compact.uniq.sort.max
     end
   end
 
-  # Filter a list of gem versions to only the latest .Z release of each .Y release
-  def latest_z_releases(versions)
-    latest = {}
-
-    versions.each do |ver|
-      major, minor, _ = ver.segments
-
-      latest[major] ||= {}
-      latest[major][minor] ||= nil
-
-      next unless latest[major][minor].nil? || ver > latest[major][minor]
-
-      latest[major][minor] = ver
-    end
-
-    latest.values.collect { |major| major.values }.flatten
-  end
-
   def ruby_for_puppet(version)
     # TODO: calculate this based on settings
     ruby_mappings = {
-      '2.5.0' => Gem::Requirement.create(['>= 6.0.0', '< 7.0.0']),
-      '2.7.0' => Gem::Requirement.create(['>= 7.0.0', '< 8.0.0']),
+      '2.7.0' => Gem::Requirement.create(['~> 7.0']),
+      '3.2.0' => Gem::Requirement.create(['~> 8.0'])
     }
 
     ruby_mappings.each do |rubyver, pup_range|
       return rubyver if pup_range.satisfied_by?(version)
     end
 
-    raise "Could not determine Ruby API version for Puppet gem version: #{version.to_s}"
+    raise "Could not determine Ruby API version for Puppet gem version: #{version}"
   end
 
   pkg.build do
     # Cache specific versions of the puppet gem
     puppet_cachedir = File.join(settings[:privatedir], 'puppet', 'ruby')
 
     gem_bins = {
-      settings[:ruby_api] => settings[:host_gem],
+      settings[:ruby_api] => settings[:host_gem]
     }
 
     bundle_bins = {
-      settings[:ruby_api] => settings[:host_bundle],
+      settings[:ruby_api] => settings[:host_bundle]
     }
 
     ruby_dirs = {
-      settings[:ruby_api] => settings[:ruby_dir],
+      settings[:ruby_api] => settings[:ruby_dir]
     }
 
-    settings[:additional_rubies]&.each do |rubyver, local_settings|
+    settings[:additional_rubies]&.each do |_rubyver, local_settings|
       gem_bins[local_settings[:ruby_api]] = local_settings[:host_gem]
       bundle_bins[local_settings[:ruby_api]] = local_settings[:host_bundle]
       ruby_dirs[local_settings[:ruby_api]] = local_settings[:ruby_dir]
     end
 
-    recent_puppets = available_puppet_gem_versions('>=6.24.0 <8.0.0')
-    latest_puppets = latest_z_releases(recent_puppets)
+    latest_puppet8 = latest_puppet_gem('~>8.0')
+    latest_puppet7 = latest_puppet_gem('~>7.0')
```

Comment:
just curious, should this be? 
```suggestion
    latest_puppet7 = latest_puppet_gem('~>7.24.0')
```

Reply:
I think this would limit us to patch releases of 7.24 where as the pessimistic dependency for 7.0 should get the latest version of 7 each time.